crates/codegen/syntax_templates/src/shared/visitor.rs

```diff
@@ -0,0 +1,80 @@
+use std::rc::Rc;
+
+use super::{cst::*, cursor::Cursor};
+
+#[allow(unused_variables)]
+pub trait Visitor<E> {
+    fn rule_enter(
+        &mut self,
+        node: &Rc<RuleNode>,
+        cursor: &Cursor,
+    ) -> Result<VisitorEntryResponse, E> {
+        Ok(VisitorEntryResponse::StepIn)
+    }
+
+    fn rule_exit(
+        &mut self,
+        node: &Rc<RuleNode>,
+        cursor: &Cursor,
```

Comment:
TOL, I wonder if we should be exposing the cursor to the visitor API, even if it cannot be used/mutated by the caller? what is the use case here?

Reply:
For two reasons
1. You can start a `Visitor` at any `Cursor` position. It is nice in the `Visitor` to be able to get to the `Cursor` and do some more internal iteration. Apart from being convenient, it makes internal and external iterators nicely symmetric.
2. Regardless of the symmetry, it is convenient in the visitor to be able to access things like the `text_range` from the `Cursor`, and also the `path_rule_nodes`. As the `Cursor` API is extended, the `Visitor` implementations can use those facilities.